### PR TITLE
PINNs solver, high-order numerical derivative

### PR DESCRIPTION
**Pull request by KirillZubov**

### CODE REVIEW

File: src/pinns_pde_solve.jl [hunk 1]
```diff
@@ -258,32 +257,37 @@ function DiffEqBase.solve(
...
         begin
-            u_in = _x[1]
-            x = collect(_x[2:end-2])
-            der_num =_x[end-1]
-            θ = _x[end]
-            (phi(x+eps_masks[dim][der_num], θ) - 2*phi(x,θ)+ phi(x-eps_masks[dim][der_num], θ) )/(ep^2)
+            if order == 1
+                return (phi(x+ε,θ) - phi(x-ε,θ))/epsilon
+            else
+                return (_derivative(x+ε,der_num,θ,order-1,ε) - _derivative(x-ε,der_num,θ,order-1,ε))/epsilon
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:284]:
```suggestion
                return (_derivative(x+ε,der_num,θ,order-1,ε) - _derivative(x-ε,der_num,θ,order-1,ε))/(2epsilon)
```

File: src/pinns_pde_solve.jl [hunk 2]
```diff
@@ -16,11 +16,19 @@ get_dict_indvars(indvars) = Dict( [Symbol(v) .=> i for (i,v) in enumerate(indvar
 
 # Calculate an order of derivative
 function count_order(_args)
-    _args[2].args[1] == :derivative ? 2 : 1
+    n = 0
+    while (_args[1] == :derivative)
+        n = n+1
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:21]:
```suggestion
        n += 1
```

File: test/NNPDE_tests.jl [hunk 3]
```diff
@@ -106,24 +106,62 @@ domains = [x ∈ IntervalDomain(0.0,2.0),
...
+prob = NeuralNetDiffEq.discretize(pde_system,discretization)
+alg = NeuralNetDiffEq.NNDE(chain,opt,autodiff=false)
+phi,res = solve(prob,alg,verbose=true, maxiters=1200)
+
+analytic_sol_func(x) = (0.0909939*x - 0.0909939)*x - 0.0322515*sin(π*x)
+xs = [domain.domain.lower:dx/10:domain.domain.upper for domain in domains][1]
+u_real  = [analytic_sol_func(x) for x in xs]
+u_predict  = [first(phi(x,res.minimizer)) for x in xs]
+
+@test u_predict ≈ u_real atol = 0.5
```
Comment (ChrisRackauckas) [inline test/NNPDE_tests.jl:163]: that seems very high

**Comment by asdasdqwdqwfsdf** [comment]
Hi @KirillZubov,

In the last version, can we now **correct** solve **8**-order numerical derivative problems, namely Dxxxxxxxx''''''''?

**Comment by ChrisRackauckas** [comment]
Yes. If you have a use case, I'd love to hear it.

**Comment by asdasdqwdqwfsdf** [comment]
@ChrisRackauckas,

how to define 8-order numerical derivative in " NeuralPDE.jl", I have tried Dxxxxxxxx''''''''", it does not work in the last version.

**Comment by ChrisRackauckas** [comment]
what did you try?

@KirillZubov I thought it was made to use FiniteDifferences.jl of any order?

**Comment by KirillZubov** [comment]
@asdasdqwdqwfsdf  You just have a typo. Used the wrong symbol " instead '.
image

**Comment by KirillZubov** [comment]
I sure, that is working well for 4-order. https://neuralpde.sciml.ai/dev/examples/pinns_example/#Example-7-:-Kuramoto%E2%80%93Sivashinsky-equation .
Increasing stability for derivative requires the use of more points stencil and the regularisation. This significantly increases the execution time of the algorithm.
It is planned to rewrite the recursive algorithm [SciML/NeuralPDE.jl#169](SciML/NeuralPDE.jl#169) and try to use some regularizations to calculate the derivative.
Does there really require a lot of attention, which equations use the derivatives of order 8+?

**Comment by KirillZubov** [comment]
```julia
@parameters x
@variables u(..)

@derivatives Dx8''''''''''~x
#@derivatives Dx''''''''"~x

eq  = Dx8(u(x)) ~ 1.
bcs = [u(0.) ~ 0.0,
       u(1.) ~ 0.0]

# Space and time domains
domains = [x ∈ IntervalDomain(0.0,1.0)]

# Discretization
dx = 0.1
# Neural network
chain = FastChain(FastDense(1,12,Flux.σ),FastDense(12,1))
discretization = NeuralPDE.PhysicsInformedNN(chain,strategy = NeuralPDE.GridTraining(dx=dx))
pde_system = PDESystem(eq,bcs,domains,[x],[u])
prob = NeuralPDE.discretize(pde_system,discretization)

res = GalacticOptim.solve(prob, ADAM(0.01); cb = cb, maxiters=2000)
phi = discretization.phi

xs = [domain.domain.lower:dx/10:domain.domain.upper for domain in domains][1]
u_predict  = [first(phi(x,res.minimizer)) for x in xs]

x_plot = collect(xs)
plot!(x_plot ,u_predict)
```

image

**Comment by asdasdqwdqwfsdf** [comment]
@ KirillZubov 

Now we may have a look at the following high order model in "NeuralPDE.jl ":

```
using NeuralPDE, Flux, ModelingToolkit, GalacticOptim, Optim, DiffEqFlux,Plots
@parameters x
@variables u(..)

@derivatives Dx6''''''~x
@derivatives Dx5'''''~x
@derivatives Dx4''''~x
@derivatives Dx3'''~x
@derivatives Dx2''~x
@derivatives Dx1'~x

eq  = Dx6(u(x))~-4.0*Dx2(u(x)) + 1.0

bcs = [Dx5(u(0)) - 1.0 ~ 0.0,
       Dx4(u(1)) ~ 0.0,
       Dx3(u(0)) ~ 0.0,
       Dx2(u(0)) ~ 1.0,
       u(0) ~ 0.0,              
       u(1.) ~ 0.0]

# Space and time domains
domains = [x ∈ IntervalDomain(0.0,1.0)]

# Discretization
dx = 0.1
# Neural network
chain = FastChain(FastDense(1,12,Flux.σ),FastDense(12,1))
discretization = NeuralPDE.PhysicsInformedNN(chain,strategy = NeuralPDE.GridTraining(dx=dx))
pde_system = PDESystem(eq,bcs,domains,[x],[u])
prob = NeuralPDE.discretize(pde_system,discretization)

cb = function (p,l)
    println("Current loss is: $l")
    return false
end

res = GalacticOptim.solve(prob, ADAM(0.01); cb = cb, maxiters=2000)
phi = discretization.phi

xs = [domain.domain.lower:dx/10:domain.domain.upper for domain in domains][1]
u_predict  = [first(phi(x,res.minimizer)) for x in xs]

x_plot = collect(xs)
plot!(x_plot ,u_predict)
```

This high order model is not able to run, because of the error: ```
ERROR: LoadError: BoundsError: attempt to access 1-element Array{Array{Float64,1},1} at index [6]
Stacktrace:
 [1] getindex at .\array.jl:809 [inlined]
...
```

P.S. We can directly find the analytical solution for this high order (6th) ODE, as we have **6 BCs** in this model.

**Comment by asdasdqwdqwfsdf** [comment]
@KirillZubov 

We can't solve your proposed high order ODE model:
```
@derivatives Dx8''''''''''~x
#@derivatives Dx''''''''"~x

eq  = Dx8(u(x)) ~ 1.
bcs = [u(0.) ~ 0.0,
       u(1.) ~ 0.0]
```
Because we need 8 BCs, not 2 BCs.

**Pull request merged**